src/tx-builder.ts

```diff
@@ -0,0 +1,314 @@
+import { execSync } from 'child_process';
+import path from 'path';
+import { bcs, BcsType } from '@mysten/bcs';
+import { SuiClient, SuiMoveNormalizedType, SuiObjectChange, SuiTransactionBlockResponseOptions } from '@mysten/sui.js/client';
+import { Keypair } from '@mysten/sui.js/dist/cjs/cryptography';
+import { TransactionBlock, TransactionObjectInput } from '@mysten/sui.js/transactions';
+import { utils as ethersUtils } from 'ethers';
+import tmp from 'tmp';
+import { updateMoveToml } from './utils';
+
+const { arrayify, hexlify } = ethersUtils;
+
+const objectCache = {} as { [id in string]: SuiObjectChange };
+
+function updateCache(objectChanges: SuiObjectChange[]) {
+    for (const change of objectChanges) {
+        if (
+            !(
+                change as {
+                    objectId: string;
+                }
+            ).objectId
+        )
+            continue;
+        objectCache[
+            (
+                change as {
+                    objectId: string;
+                }
+            ).objectId
+        ] = change;
+    }
+}
+
+function getObject(tx: TransactionBlock, object: TransactionObjectInput) {
+    if (Array.isArray(object)) {
+        object = hexlify(object);
+    }
+
+    if (typeof object === 'string') {
+        const cached = objectCache[object];
+
+        if (cached) {
+            // TODO: figure out how to load the object version/digest into the TransactionBlock because it seems impossible for non gas payment objects
+            const txObject = tx.object(object);
+            return txObject;
+        }
+
+        return tx.object(object);
+    }
+
+    return object;
+}
+
+function getTypeName(type: SuiMoveNormalizedType): string {
+    type Type = { address: string; module: string; name: string; typeArguments: string[] };
+
+    function get(type: Type) {
+        let name = `${type.address}::${type.module}::${type.name}`;
+
+        if (type.typeArguments.length > 0) {
+            name += `<${type.typeArguments[0]}`;
+
+            for (let i = 1; i < type.typeArguments.length; i++) {
+                name += `,${type.typeArguments[i]}`;
+            }
+
+            name += '>';
+        }
+
+        return name;
+    }
+
+    if ((type as { Struct: Type }).Struct) {
+        return get((type as { Struct: Type }).Struct);
+    } else if ((type as { Reference: SuiMoveNormalizedType }).Reference) {
+        return getTypeName((type as { Reference: SuiMoveNormalizedType }).Reference);
+    } else if ((type as { MutableReference: SuiMoveNormalizedType }).MutableReference) {
+        return getTypeName((type as { MutableReference: SuiMoveNormalizedType }).MutableReference);
+    } else if ((type as { Vector: SuiMoveNormalizedType }).Vector) {
+        return `vector<${getTypeName((type as { Vector: SuiMoveNormalizedType }).Vector)}>`;
+    }
```

Comment:
can you pls try using [this strategy](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-in-operator-narrowing) for narrowing the type? other strategies in the link can be useful in other places as well since sui has many union types in their sdk.